rustls/src/msgs/handshake.rs

```diff
@@ -2532,39 +2552,109 @@ impl Codec<'_> for EchConfigContents {
                     .map_err(|_| InvalidMessage::InvalidServerName)?
                     .to_owned()
             },
-            extensions: PayloadU16::read(r)?,
+            extensions: Vec::read(r)?,
         })
     }
 }
 
+/// An encrypted client hello (ECH) config.
 #[derive(Clone, Debug, PartialEq)]
-pub struct EchConfig {
-    pub version: EchVersion,
-    pub contents: EchConfigContents,
+pub enum EchConfig {
+    /// A recognized V18 ECH configuration.
+    V18(EchConfigContents),
+    /// An unknown version ECH configuration.
+    Unknown {
+        version: EchVersion,
+        contents: PayloadU16,
+    },
+}
+
+impl TlsListElement for EchConfig {
+    const SIZE_LEN: ListLength = ListLength::U16;
 }
 
 impl Codec<'_> for EchConfig {
     fn encode(&self, bytes: &mut Vec<u8>) {
-        self.version.encode(bytes);
-        let mut contents = Vec::with_capacity(128);
-        self.contents.encode(&mut contents);
-        let length: &mut [u8; 2] = &mut [0, 0];
-        codec::put_u16(contents.len() as u16, length);
-        bytes.extend_from_slice(length);
-        bytes.extend(contents);
+        match self {
+            Self::V18(c) => {
+                // Encode the contents
+                let mut contents = Vec::with_capacity(128);
+                c.encode(&mut contents);
+
+                // Encode the length of the contents.
+                let length: &mut [u8; 2] = &mut [0, 0];
+                codec::put_u16(contents.len() as u16, length);
+
+                // Write the version, the length, and the contents.
+                EchVersion::V18.encode(bytes);
+                bytes.extend_from_slice(length);
+                bytes.extend(contents);
+            }
+            Self::Unknown { version, contents } => {
+                // Unknown configuration versions are opaque.
+                version.encode(bytes);
+                contents.encode(bytes);
+            }
+        }
     }
 
     fn read(r: &mut Reader) -> Result<Self, InvalidMessage> {
         let version = EchVersion::read(r)?;
         let length = u16::read(r)?;
-        Ok(Self {
-            version,
-            contents: EchConfigContents::read(&mut r.sub(length as usize)?)?,
+
+        Ok(match version {
+            EchVersion::V18 => Self::V18(EchConfigContents::read(&mut r.sub(length as usize)?)?),
```

Comment:
any value in extracting `r.sub(length as usize)?` -- seems like every match arm here must do that to be correct.

Reply:
Done.